### ISSUE TITLE
Plan: Comprehensive C++ code generation from the PyTorch C++ docs

### ISSUE THREAD

**Issue by drzo** [ReZorg/pytorchcppdocs#2]
<div class="markdown-body MarkdownRenderer-module__container__sJ8jf" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><h1>Plan: Comprehensive C++ code generation from the PyTorch C++ docs</h1>
<h2>Goal</h2>
<p>Generate a buildable, header-organized C++ codebase that mirrors the documented PyTorch C++ API — a complete set of compilable example/usage files (and where appropriate API surface stubs) for every documented feature, using the correct C++ file type per role.</p>
<h2>Source-of-truth inputs (in priority order)</h2>
<ol>
<li node="[object Object]"><code node="[object Object]">llms-full.txt</code> — the full flattened doc corpus; cheapest to parse programmatically.</li>
<li node="[object Object]"><code node="[object Object]">_sources/**/*.md.txt</code> — structured per-topic specs with explicit directives and header lists.</li>
<li node="[object Object]"><code node="[object Object]">_coverage/**</code> — the authoritative list of real header paths to mirror.</li>
<li node="[object Object]"><code node="[object Object]">api/**/*.html</code> — only for cross-checking rendered signatures; not a primary input.</li>
</ol>
<h2>File-type convention (per role)</h2>
<ul>
<li node="[object Object]"><code node="[object Object]">.h</code> — public API surface stubs, umbrella headers, and the stable-ABI / C-compatible interface (<code node="[object Object]">torch/csrc/stable/**</code>).</li>
<li node="[object Object]"><code node="[object Object]">.hpp</code> — template-heavy and header-only implementations: <code node="[object Object]">torch::nn</code> module templates (<code node="[object Object]">Cloneable</code>, <code node="[object Object]">ModuleHolder</code>/pimpl), <code node="[object Object]">torch::data</code> datasets/samplers/transforms (all templated), <code node="[object Object]">torch::OrderedDict</code>, options structs, and inline functional wrappers.</li>
<li node="[object Object]"><code node="[object Object]">.cpp</code> — non-template implementations and every compilable example program (each with <code node="[object Object]">main()</code>).</li>
<li node="[object Object]"><code node="[object Object]">.cu</code> — CUDA-only kernels/examples (guards, streams, device kernels) gated behind <code node="[object Object]">C10_CUDA_ENABLED</code>/NVCC.</li>
<li node="[object Object]"><code node="[object Object]">.hh</code> / <code node="[object Object]">.c</code> — <strong>not generated</strong>: not used anywhere in the PyTorch C++ API; document this as a deliberate exclusion.</li>
</ul>
<h2>Target layout</h2>
<figure class="CodeBlock-module__container___ry3V" aria-labelledby="_r_1l_"><div class="CodeBlock-module__header__tJzrz"><span class="LanguageDot-module__languageDot__WZEoH"></span><span id="_r_1l_" class="CodeBlock-module__languageName__iQIeg">Code</span></div><div class="CodeBlock-module__copyContainer__eNjNy"><div class="CodeBlock-module__copyContent__qV51h"><button data-component="IconButton" type="button" class="prc-Button-ButtonBase-9n-Xk CodeBlock-module__copyButton__buwCD prc-Button-IconButton-fyge7" data-loading="false" data-no-visuals="true" data-size="medium" data-variant="invisible" aria-labelledby="_r_1m_"><svg data-component="Octicon" aria-hidden="true" focusable="false" class="octicon octicon-copy" viewBox="0 0 16 16" width="16" height="16" fill="currentColor" display="inline-block" overflow="visible" style="vertical-align: text-bottom;"><path d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z"></path><path d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z"></path></svg></button><span class="CopyToClipboardIconButton-module__tooltip__WyiwL prc-TooltipV2-Tooltip-tLeuB" data-direction="s" data-component="Tooltip" aria-label="Copy code" aria-hidden="true" id="_r_1m_" popover="auto">Copy code</span></div></div><div class="CodeBlock-module__codeContainer___2Otq"><pre class="CodeBlock-module__code__N23o9" tabindex="0"><code class="">generated/
  CMakeLists.txt            # root build, find_package(Torch), mirrors examples/ pattern
  README.md                 # generation provenance + build instructions
  include/                  # API surface stubs, mirrors _coverage header tree
    at/  c10/  torch/ ...   # .h / .hpp per convention above
  src/                      # non-template implementations (.cpp) for stubs
  examples/&lt;module&gt;/        # one buildable .cpp example per doc page (63 pages)
</code></pre></div></figure>
<p>Header paths under <code node="[object Object]">include/</code> replicate the real include paths listed in <code node="[object Object]">_coverage</code> (e.g. <code node="[object Object]">torch/nn/modules/conv.h</code>, <code node="[object Object]">c10/core/Device.h</code>) so examples use authentic <code node="[object Object]">#include</code> lines.</p>
<h2>Work streams (by doc module)</h2>
<ol>
<li node="[object Object]"><strong>ATen</strong> (tensor, creation, indexing, accessors) — <code node="[object Object]">at::Tensor</code>, <code node="[object Object]">TensorOptions</code>, <code node="[object Object]">Scalar</code>, <code node="[object Object]">ScalarType</code>, <code node="[object Object]">Layout</code>, <code node="[object Object]">DeviceGuard</code>; factory examples (<code node="[object Object]">ones/randn/zeros</code>), accessor/iterator examples. → <code node="[object Object]">include/ATen/*.h</code> + <code node="[object Object]">examples/aten/*.cpp</code>.</li>
<li node="[object Object]"><strong>c10</strong> (device, guards, streams, types, utilities) — <code node="[object Object]">Device</code>, <code node="[object Object]">DeviceType</code>, <code node="[object Object]">Stream</code>, guards, <code node="[object Object]">ArrayRef</code>, <code node="[object Object]">Optional</code>. → <code node="[object Object]">include/c10/**</code> + examples.</li>
<li node="[object Object]"><strong>Autograd</strong> (gradient, custom_functions, modes) — <code node="[object Object]">torch::autograd</code>, <code node="[object Object]">Function</code> subclass example, <code node="[object Object]">GradMode</code>/<code node="[object Object]">NoGradGuard</code>. → examples + minimal stubs.</li>
<li node="[object Object]"><strong>nn</strong> (all 14 pages) — module stubs (<code node="[object Object]">Conv1/2/3d</code>, <code node="[object Object]">ConvTranspose*</code>, <code node="[object Object]">Linear</code>, <code node="[object Object]">Embedding</code>, RNN/LSTM/GRU, Transformer, pooling, normalization, dropout, activation, loss), containers (<code node="[object Object]">Sequential</code>, <code node="[object Object]">ModuleList</code>, <code node="[object Object]">ModuleDict</code>), the <code node="[object Object]">functional</code> stateless API, options structs, <code node="[object Object]">pimpl</code>/<code node="[object Object]">Cloneable</code> patterns. These are template/macro-heavy → mostly <code node="[object Object]">.hpp</code> plus per-feature <code node="[object Object]">.cpp</code> examples (model definitions + <code node="[object Object]">forward</code>).</li>
<li node="[object Object]"><strong>optim</strong> (gradient_descent, adaptive, second_order, schedulers) — <code node="[object Object]">SGD/Adam/AdamW/RMSprop/Adagrad/LBFGS</code>, options, param groups, LR schedulers. → stubs + training-step examples.</li>
<li node="[object Object]"><strong>data</strong> (dataloader, datasets, samplers, transforms) — <code node="[object Object]">Dataset</code>/<code node="[object Object]">MapDataset</code>/<code node="[object Object]">ChunkDataset</code>, <code node="[object Object]">MNIST</code>, samplers, transforms, <code node="[object Object]">DataLoader</code> options; custom-dataset example. → <code node="[object Object]">.hpp</code> templates + examples.</li>
<li node="[object Object]"><strong>serialize</strong> (save_load, archives, checkpoints) — <code node="[object Object]">torch::save/load</code>, <code node="[object Object]">InputArchive</code>/<code node="[object Object]">OutputArchive</code>, checkpoint patterns. → stubs + examples.</li>
<li node="[object Object]"><strong>library</strong> (registration, custom_classes, versioning) — <code node="[object Object]">TORCH_LIBRARY</code>, <code node="[object Object]">TORCH_LIBRARY_IMPL</code>, <code node="[object Object]">TORCH_LIBRARY_FRAGMENT</code>, <code node="[object Object]">Library</code>, <code node="[object Object]">CppFunction</code>, custom class binding. → registration <code node="[object Object]">.cpp</code> examples.</li>
<li node="[object Object]"><strong>stable</strong> (registration, operators, utilities) — the stable-ABI C-style surface (<code node="[object Object]">STABLE_TORCH_LIBRARY</code>, <code node="[object Object]">torch::stable::*</code>, headeronly types). → <code node="[object Object]">.h</code> stubs + examples; this is the one place <code node="[object Object]">.h</code> C-compatible style is mandatory.</li>
<li node="[object Object]"><strong>cuda / xpu</strong> (streams, guards, utilities) — device guard/stream examples; CUDA kernel <code node="[object Object]">.cu</code> example gated by availability.</li>
<li node="[object Object]"><strong>Top-level guides</strong> (index, frontend end-to-end, installing) — the frontend end-to-end example becomes a full <code node="[object Object]">examples/frontend/</code> training app; <code node="[object Object]">examples/minimal</code>/<code node="[object Object]">mnist</code> are kept as reference and their CMake pattern reused.</li>
</ol>
<h2>Generation approach</h2>
<ul>
<li node="[object Object]"><strong>Parse, don't hand-copy</strong>: write a small Python generator that (a) walks <code node="[object Object]">_sources/**/*.md.txt</code>, extracts every <code node="[object Object]">```cpp</code> block, every <code node="[object Object]">doxygen*</code>/<code node="[object Object]">cpp:*</code> symbol directive, and every "Header Files" entry; (b) groups symbols by namespace/module; (c) emits the header stubs and example files from templates. <code node="[object Object]">llms-full.txt</code> is the fallback corpus for anything missed.</li>
<li node="[object Object]"><strong>Coverage-driven completeness check</strong>: diff the generated <code node="[object Object]">include/</code> tree against the 149 header paths in <code node="[object Object]">_coverage/</code> to prove every documented header is represented; emit a coverage report.</li>
<li node="[object Object]"><strong>Directives → code mapping</strong>: <code node="[object Object]">doxygenclass X</code> → a class stub + usage example; <code node="[object Object]">doxygenfunction</code> → a callable example; <code node="[object Object]">doxygendefine</code>/<code node="[object Object]">TORCH_LIBRARY*</code> → registration example; <code node="[object Object]">cpp:enum-class</code>/<code node="[object Object]">enumerator</code> → a typed enum stub; <code node="[object Object]">cpp:function</code> → method on the owning class stub.</li>
<li node="[object Object]"><strong>Each example is self-contained</strong> with a <code node="[object Object]">main()</code> and mirrors the doc's canonical snippet, so the entire <code node="[object Object]">examples/</code> tree doubles as compile-tests.</li>
</ul>
<h2>Build &amp; validation</h2>
<ul>
<li node="[object Object]">Root <code node="[object Object]">CMakeLists.txt</code> using the documented LibTorch pattern (<code node="[object Object]">find_package(Torch REQUIRED)</code>, <code node="[object Object]">TORCH_CXX_FLAGS</code>, <code node="[object Object]">CXX_STANDARD 20</code>, MSVC DLL copy block), one <code node="[object Object]">add_executable</code> per example, optional <code node="[object Object]">add_library</code> for stubs.</li>
<li node="[object Object]">CI-style validation script: configure + build every target against a LibTorch install; treat each compiled example as a passing test. CUDA <code node="[object Object]">.cu</code> targets build only when <code node="[object Object]">CUDAToolkit</code> is found.</li>
<li node="[object Object]">Lint/format with clang-format using the existing examples' style.</li>
</ul>
<h2>Sequencing</h2>
<ol>
<li node="[object Object]">Generator + file-type templates + root CMake scaffolding.</li>
<li node="[object Object]">ATen + c10 + autograd (foundational).</li>
<li node="[object Object]">nn (largest surface) + functional.</li>
<li node="[object Object]">optim + data + serialize (frontend training path).</li>
<li node="[object Object]">library + stable (extension/ABI surface).</li>
<li node="[object Object]">cuda/xpu + top-level guide apps.</li>
<li node="[object Object]">Coverage report vs <code node="[object Object]">_coverage</code>, full build, docs (<code node="[object Object]">generated/README.md</code>).</li>
</ol>
<h2>Notes / constraints</h2>
<ul>
<li node="[object Object]">This repo is auto-generated from <code node="[object Object]">pytorch/pytorch/docs/cpp/source</code>; the generator should live in a self-contained <code node="[object Object]">generated/</code> (or a <code node="[object Object]">codegen/</code> tool dir) so it can be re-run, and the provenance note should state that doc-source edits belong upstream.</li>
<li node="[object Object]">No <code node="[object Object]">.hh</code> or <code node="[object Object]">.c</code> files are produced; <code node="[object Object]">.h</code> is reserved for C-compatible/stable-ABI headers, <code node="[object Object]">.hpp</code> for C++ template/header-only code, <code node="[object Object]">.cpp</code> for implementations and examples, <code node="[object Object]">.cu</code> for CUDA.</li>
</ul></div>
